<a href="https://colab.research.google.com/github/Ysagi2000/skillfactory/blob/main/%D0%9C%D0%90%D0%A2%D0%A1%D0%A2%D0%90%D0%A2_%D0%BC%D0%BE%D0%B4%D1%83%D0%BB%D1%8C_4_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Задание 1 (на сопоставление)

In [4]:
"Ответ: 1 - E, 2 - В, 3 - F, 4 - G, 5 - A"

'ответ: 1 - E, 2 - В, 3 - F, 4 - G, 5 - A'

Задание 2 (про порошок)

Для проверки гипотезы о среднем значении используем однообразный z-тест.

**Нулевая гипотеза** (H0):
μ = 3.00 кг

**Альтернативная гипотеза** (H1):
μ ≠ 3.00 кг

In [5]:
import numpy as np
from scipy import stats

sample_mean = 3.005
population_mean = 3
population_std = 0.015
sample_size = 50
alpha = 0.05
# Расчет z-оценки
z_score = (sample_mean-population_mean)/(population_std/np.sqrt(50))
print('z-оценка :',z_score)

# Принятие решения на основе критического значения z
# Расчет критической z-оценки
z_critical = stats.norm.ppf(1-alpha/2)
print('Критическая z-оценка :',z_critical)
if z_score >  z_critical:
    print("Отклонить нулевую гипотезу")
else:
    print("Нельзя отклонить нулевую гипотезу")

# Принятие решения на основе p-значения
# P-Value : вероятность получить значение меньше, чем z-оценка
p_value = 1-stats.norm.cdf(z_score)
print('p-value :',p_value)
if p_value <  alpha:
    print("Отклонить нулевую гипотезу")
else:
    print("Нельзя отклонить нулевую гипотезу")

z-оценка : 2.3570226039551083
Критическая z-оценка : 1.959963984540054
Отклонить нулевую гипотезу
p-value : 0.009211062727050745
Отклонить нулевую гипотезу


**Выводы:**
Поскольку
z ≈ 2.36, а 2.36 больше чем 1.96, мы отклоняем нулевую гипотезу.

Таким образом, образец предоставляет достаточные доказательства для того, чтобы утверждать, что истинное среднее значение не равно 3.00 кг, и, следовательно, дозатор требует настройки.

Задание 3 (про двигатели)

Для того чтобы проверить, есть ли статистически значимая разница между средними прочностными характеристиками двух веществ, можно использовать t-тест для независимых выборок.

У нас есть следующие данные:

**Для вещества 1:**

Размер выборки (n1) = 15
Средняя прочность (M1) = 90 кг*мм⁻²
Стандартное отклонение (SD1) = 2.3 кг*мм⁻²

**Для вещества 2:**

Размер выборки (n2) = 10
Средняя прочность (M2) = 88 кг*мм⁻²
Стандартное отклонение (SD2) = 2.3 кг*мм⁻²

**Исходные гипотезы:**

H0: Средние прочностные характеристики не отличаются (M1 = M2)

H1: Средние прочностные характеристики отличаются (M1 ≠ M2)

In [12]:
num = 15
num_mean = 90
num_std = 2.3
num_two = 10
num_two_mean = 88
num_two_std = 2.2
alpha = 0.05
# найдем несмещенные стандартные отклонения
num_std_correct = num_std * np.sqrt(num/(num-1))
n2_std_corr = num_two_std * np.sqrt(num_two/(num_two-1))

# посчитаем статистики, с учетом неравенства дисперсий
t_score, p_value = stats.ttest_ind_from_stats(num_mean, num_std_correct, num, num_two_mean, num_std_correct, num_two, equal_var=False)

# рассчитаем приведенное количество степеней свободы
v = (num_std**2/num + num_two_std**2/num_two)**2/(num_std**4/(num**2*(num-1)) + num_two_std**4/(num_two**2*(num_two-1)))

# для данного количества степеней свободы рассчитаем критическое значение статистики
t_critical = stats.t.ppf(1 - alpha/2, df = v)

if np.abs(t_score) > t_critical:
    print("Отклонить нулевую гипотезу")
else:
    print("Нельзя отклонить нулевую гипотезу")

if p_value < alpha:
    print("Отклонить нулевую гипотезу")
else:
    print("Нельзя отклонить нулевую гипотезу")
print(f" {t_score:.4f}\n {t_critical:.4f}\n {p_value:.4f}\n")

Нельзя отклонить нулевую гипотезу
Нельзя отклонить нулевую гипотезу
 2.0578
 2.0856
 0.0533



Сравнение с критическим значением:
на уровне значимости5% (уровне доверия 95%) нет оснований отвергнуть нулевую гипотезу в пользу альтернативы.

Критическое значение для t с 23 df примерно равно 2.069.

**Заключение**:
Так как рассчитанная t-статистика (≈ 2.14) больше критического значения (≈ 2.069), мы отклоняем нулевую гипотезу.

Таким образом, можно сделать вывод, что на уровне значимости 5% существует статистически значимая разница между средними прочностными характеристиками веществ.

Задание 4 (про токи)

In [27]:

x1 = np.array([80.1, 82.3, 84.1, 82.6, 85.3, 81.3, 83.2, 81.7, 82.2, 81.4])
x2 = np.array([80.7, 81.3, 84.6, 81.7, 86.3, 84.3, 83.7, 84.7, 82.8, 84.4, 85.2, 84.9])

n1 = len(x1)
n2 = len(x2)
alpha = 0.05

# посчитаем статистики
res = stats.ttest_ind(x1, x2)
t_score = res[0]
p_value = res[1]

# рассчитаем критическое значение статистики
t_critical = stats.t.ppf(1 - alpha/2, df = res.df)

if np.abs(t_score) > t_critical:
    print("Отклонить нулевую гипотезу")
else:
    print("Нельзя отклонить нулевую гипотезу")

if p_value < alpha:
    print("Отклонить нулевую гипотезу")
else:
    print("Нельзя отклонить нулевую гипотезу")
print(f" {t_score:.4f},\n {t_critical:.4f},\n {p_value:.4f},\n {int(res.df)}")

Нельзя отклонить нулевую гипотезу
Нельзя отклонить нулевую гипотезу
 -1.8657,
 2.0860,
 0.0768,
 20


**Статистический вывод:**

На имеющихся данных, на уровне значимости 5% (уровне доверия 95%) нет оснований отвергнуть нулевую гипотезу в пользу альтернативы.

**Содержательный вывод**:

Оснований не заменять старую конструкцию схемы на новую не найдено.

Задание 5 (про глюкозу)

**Шаги для анализа**

**Расчет разностей:** Вычислим разности между уровнями глюкозы "до" и "после" приема препарата.

**Проведение парного t-теста:** Сравним среднее значение разностей с нулем с помощью t-теста.

**Интерпретация результата:** Сравним p-значение с уровнем значимости 0.05.

In [33]:
import numpy as np
from scipy import stats

# Данные по уровням глюкозы (до и после препарата)
before = np.array([9.6, 8.1, 8.8, 7.9, 9.2, 8.0, 8.4, 10.1, 7.8, 8.1])
after = np.array([7.7, 6.2, 7.4, 7.5, 8.3, 6.2, 8.1, 8.9, 7.4, 7.0])

# Выполнение парного t-теста
t_statistic, p_value = stats.ttest_rel(before, after)

# Уровень значимости
alpha = 0.05

# Проверка гипотезы
if p_value < alpha:
    conclusion = "Мы отклоняем нулевую гипотезу: препарат имеет статистически значимый эффект на уровень глюкозы."
else:
    conclusion = "Мы не отклоняем нулевую гипотезу: нет статистически значимого эффекта от препарата."

# Вывод результатов
print(f"T-статистика: {t_statistic:.4f}")
print(f"P-значение: {p_value:.4f}")
print(conclusion)


T-статистика: 5.7130
P-значение: 0.0003
Мы отклоняем нулевую гипотезу: препарат имеет статистически значимый эффект на уровень глюкозы.


**Статистический вывод**

На имеющихся данных, на уровне значимости 5% (уровне доверия 95%) есть основания отвергнуть нулевую гипотезу в пользу альтернативы.

Содержательный вывод: проведенные исследования дают основания считать средство эффективным.


**Содержательные выводы**

Так как p-значение < 0.05, то считается, что препарат имеет статистически значимый эффект на уровень глюкозы. другими словами, средство эффективно.

In [ ]:
'''Ваш ответ на задание №5
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

Задание 6 (про тросы)

Мы выбираем двухвыборочный t-тест Уэлча (t-тест для случаев, когда дисперсии не равны) по следующим причинам:

Мы сравниваем две независимые группы данных.
Указано, что дисперсии в этих группах различаются.
Данные представляют собой количественные измерения.
Мы предполагаем, что данные нормально распределены.

In [35]:
import numpy as np
from scipy import stats

# Данные для анализа
modern_cable = np.array([92.7, 91.6, 94.7, 93.7, 96.5, 94.3, 93.7, 96.8, 98.9, 99.9])
classic_cable = np.array([90.1, 92.2, 94.3, 92.1, 93.4, 91.1, 93.2, 91.5])

# Вычисление основных статистиков
modern_mean = np.mean(modern_cable)
classic_mean = np.mean(classic_cable)
modern_std = np.std(modern_cable, ddof=1)
classic_std = np.std(classic_cable, ddof=1)

# Проведение одностороннего t-теста Уэлча
t_statistic, p_val = stats.ttest_ind(modern_cable, classic_cable, equal_var=False)

# Приведение p-value к одностороннему
p_val_one_sided = p_val / 2  # делим на 2 для правостороннего теста

# Вывод статистических показателей
print("Статистические результаты:")
print(f"Среднее значение современного кабеля: {modern_mean:.2f}")
print(f"Среднее значение классического кабеля: {classic_mean:.2f}")
print(f"Стандартное отклонение современного кабеля: {modern_std:.2f}")
print(f"Стандартное отклонение классического кабеля: {classic_std:.2f}")
print(f"\nРезультаты t-теста:")
print(f"t-статистика: {t_statistic:.4f}")
print(f"p-value (односторонний тест): {p_val_one_sided:.4f}")

# Вывод результатов теста
print("\nЗаключение:")
if p_val_one_sided < 0.01:
    print("Рекомендуется замена классического кабеля на современный (имеется статистически значимое улучшение)")
else:
    print("Нет достаточных оснований для замены кабеля")


Статистические результаты:
Среднее значение современного кабеля: 95.28
Среднее значение классического кабеля: 92.24
Стандартное отклонение современного кабеля: 2.68
Стандартное отклонение классического кабеля: 1.36

Результаты t-теста:
t-статистика: 3.1204
p-value (односторонний тест): 0.0038

Заключение:
Рекомендуется замена классического кабеля на современный (имеется статистически значимое улучшение)


Задание 7 (про абитуриентов)

In [37]:
import numpy as np
from scipy import stats

# Исходные данные
group1 = np.array([7, 6, 5, 4, 6, 2, 10, 3, 8, 5])
group2 = np.array([9, 6, 4, 5, 7, 4, 10, 6, 9, 6])
alpha_level = 0.01

# Проведение парного t-теста
test_result = stats.ttest_rel(group1, group2, alternative='less')

# Получение статистики теста
t_statistic = test_result.statistic
p_value = test_result.pvalue
degrees_of_freedom = test_result.df

# Вычисляем критическое значение t
critical_t_value = stats.t.ppf(1 - alpha_level, df=degrees_of_freedom)

# 1. Анализ по критическому значению
print("Анализ по критическому значению:")
if t_statistic < critical_t_value:  # Для одностороннего теста less
    print("Отклоняем нулевую гипотезу H₀")
else:
    print("Нет оснований для отклонения нулевой гипотезы H₀")

# 2. Анализ по p-value
print("\nАнализ по p-value:")
if p_value < alpha_level:
    print("Отклоняем нулевую гипотезу H₀")
else:
    print("Нет оснований для отклонения нулевой гипотезы H₀")

# Вывод значений t-статистики и p-value
print("\nРезультаты теста:")
print(f"t-статистика: {t_statistic:.4f}")
print(f"Критическое значение t: {critical_t_value:.4f}")
print(f"p-значение: {p_value:.4f}")
print(f"Степени свободы: {degrees_of_freedom}")


Анализ по критическому значению:
Отклоняем нулевую гипотезу H₀

Анализ по p-value:
Нет оснований для отклонения нулевой гипотезы H₀

Результаты теста:
t-статистика: -2.7386
Критическое значение t: 2.8214
p-значение: 0.0114
Степени свободы: 9


**Статистический вывод**: на имеющихся данных, на уровне значимости 1% (уровне доверия 99%) нет оснований отвергнуть нулевую гипотезу в пользу альтернативы.

**Содержательный вывод:** подтверждения, что подготовительные курсы положительно влияют на результаты абитуриентов, не получено.

Задание 8 (про фрукты и IQ)

In [38]:
import numpy as np
from scipy import stats

# Данные
data1 = np.array([125, 138, 112, 130, 124, 125, 110, 112, 117, 126])
data2 = np.array([124, 102, 142, 130, 131, 124, 127, 115, 108, 109])
alpha_level = 0.05
sample_size = len(data1)

# Вычисление выборочных дисперсий
variance1 = data1.var(ddof=1)
variance2 = data2.var(ddof=1)

print(f'Выборочная дисперсия S1: {variance1:.2f}, S2: {variance2:.2f}')

# Вычисление F-статистики
F_statistic = max(variance1, variance2) / min(variance1, variance2)
print(f'F-статистика: {F_statistic:.4f}')

# Поиск критического значения F
F_critical = stats.f.ppf(1 - alpha_level, dfn=sample_size - 1, dfd=sample_size - 1)
print(f'Критическое значение F: {F_critical:.4f}')

# Принятие решения на основе критического значения
if F_statistic > F_critical:
    print("Отклоняем нулевую гипотезу о равенстве дисперсий")
else:
    print("Нет оснований для отклонения нулевой гипотезы о равенстве дисперсий")


Выборочная дисперсия S1: 80.77, S2: 153.96
F-статистика: 1.9062
Критическое значение F: 3.1789
Нет оснований для отклонения нулевой гипотезы о равенстве дисперсий


In [ ]:
'''Ваш ответ на задание №8
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

**Статистический вывод**: на имеющихся данных, на уровне значимости 5% (уровне доверия 95%) нет оснований отвергнуть нулевую гипотезу в пользу альтернативы.

**Содержательный вывод:** подтверждения, что дисперсии двух групп отличаются, не получено.

Задание 9 (про самолеты)

In [39]:
import numpy as np
from scipy import stats

# Наблюдаемые данные о вылетах и количестве успешных попыток
X = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9])
Y = np.array([36, 23, 11, 13, 8, 3, 2, 1, 3])
alpha = 0.05

# Сумма наблюдений
n = Y.sum()
print(f'Сумма наблюдений (n): {n}')

# Оценочная вероятность
p_curr = n / np.dot(X, Y)
print(f'Оценочная вероятность (p): {p_curr:.4f}')

# Вычисление ожидаемых значений
Y_p = np.zeros(len(Y))
for i in range(len(Y)):
    Y_p[i] = stats.geom.pmf(X[i], p_curr) * n

# Коррекция для соблюдения условия суммы
dev = (n - Y_p.sum()) / len(Y)
Y_p += dev

# Тест Хи-квадрат
chi2_score, p_value = stats.chisquare(Y, Y_p, ddof=1)

print(f'Chi-squared statistic: {chi2_score:.4f}, p-value: {p_value:.4f}')

# Степени свободы
nu = Y.shape[0] - 1 - 1  # k - 1 - m, где k - количество категорий, m - количество параметров
print(f'Степени свободы (nu): {nu}')

# Критическое значение для теста
chi2_critical = stats.chi2.ppf(1 - alpha, df=nu)
print(f'Критическое значение Хи-квадрат: {chi2_critical:.4f}')

# Решение на основе критического значения
if chi2_score > chi2_critical:
    print("Отклонить нулевую гипотезу о соответствии распределений")
else:
    print("Нет оснований для отклонения нулевой гипотезы")

# Решение на основе p-значения
if p_value < alpha:
    print("Отклонить нулевую гипотезу о соответствии распределений")
else:
    print("Нет оснований для отклонения нулевой гипотезы")


Сумма наблюдений (n): 100
Оценочная вероятность (p): 0.3650
Chi-squared statistic: 6.4975, p-value: 0.4830
Степени свободы (nu): 7
Критическое значение Хи-квадрат: 14.0671
Нет оснований для отклонения нулевой гипотезы
Нет оснований для отклонения нулевой гипотезы


**Статистический вывод:** на имеющихся данных, на уровне значимости 5% (уровне доверия 95%) нет оснований отвергнуть нулевую гипотезу в пользу альтернативы.

**Содержательный вывод:** нет оснований считать, что данные распределены не геометрически.

In [ ]:
'''Ваш ответ на задание №9
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

Задание 10 (про цвет глаз и волос)

In [40]:
import numpy as np
from scipy import stats

# Определение таблицы контингентности
contingency_table = np.array([[35, 52, 13],
                               [28, 42, 10],
                               [7, 10, 3]])

# Проведение теста Хи-квадрат
chi2_statistic, p_value, _, _ = stats.chi2_contingency(contingency_table)

# Вычисление степеней свободы
degrees_of_freedom = (contingency_table.shape[0] - 1) * (contingency_table.shape[1] - 1)
alpha_level = 0.05

# Определение критического значения Хи-квадрат
chi2_critical_value = stats.chi2.ppf(1 - alpha_level, df=degrees_of_freedom)

# Вывод значений статистики и критического значения
print(f"Значение статистики Хи-квадрат: {chi2_statistic:.4f}")
print(f"Критическое значение Хи-квадрат: {chi2_critical_value:.4f}")

# Принятие решения на основе критического значения
if chi2_statistic > chi2_critical_value:
    print("Решение: Отклоняем нулевую гипотезу.")
else:
    print("Решение: Нельзя отклонить нулевую гипотезу.")

# Вывод p-значения
print(f"p-значение: {p_value:.4f}")

# Принятие решения на основе p-значения
if p_value < alpha_level:
    print("Решение: Отклоняем нулевую гипотезу на основе p-значения.")
else:
    print("Решение: Нельзя отклонить нулевую гипотезу на основе p-значения.")


Значение статистики Хи-квадрат: 0.0962
Критическое значение Хи-квадрат: 9.4877
Решение: Нельзя отклонить нулевую гипотезу.
p-значение: 0.9989
Решение: Нельзя отклонить нулевую гипотезу на основе p-значения.


**Статистический вывод**: на имеющихся данных, на уровне значимости 5% (уровне доверия 95%) нет оснований отвергнуть нулевую гипотезу в пользу альтернативы.

**Содержательный вывод:** нет оснований считать, что существует связь между цветом волос и цветом глаз.

Задание 11 (про тигров)

**Цель исследования**

Задача заключается в сравнении четырех групп тигров в зависимости от типа корма, чтобы проанализировать, как этот фактор влияет на их вес. В этом случае уместно использовать однофакторный ANOVA.

**Гипотезы**

**Нулевая гипотеза (H0)**: Средние веса тигров в каждой из четырех групп, получающих разные корма, одинаковы.

**Альтернативная гипотеза (H1)**: Существуют различия в среднем весе тигров в зависимости от

In [2]:
X1 = np.array([151, 135, 137, 118, 132, 135, 131, 121, 137, 140, 152, 133, 151, 132])
X2 = np.array([108, 123, 145, 103, 110, 124, 116, 108, 104, 130, 106, 105, 114, 108])
X3 = np.array([140, 152, 149, 132, 124, 146, 156, 137, 124, 145, 147, 139, 157, 139])
X4 = np.array([127, 135, 110, 116, 119, 123, 102, 121, 118, 120, 114, 134, 121, 109])
res = stats.f_oneway(X1, X2, X3, X4)
alpha = 0.05
if res.pvalue < alpha:
    print("Отклонить нулевую гипотезу")
else:
    print("Нельзя отклонить нулевую гипотезу")


Отклонить нулевую гипотезу


Результаты показывают, что нулевая гипотеза была отклонена. Это указывает на наличие статистически значимых различий в весе тигров в зависимости от типа корма на уровне значимости 5% (доверительный уровень 95%).

Поскольку ANOVA показал значимость, был применен тест Тьюки для более детального анализа различий между группами:

In [3]:
t_res = stats.tukey_hsd(X1, X2, X3, X4)
print(t_res)


Tukey's HSD Pairwise Group Comparisons (95.0% Confidence Interval)
Comparison  Statistic  p-value  Lower CI  Upper CI
 (0 - 1)     21.500     0.000    10.981    32.019
 (0 - 2)     -5.857     0.458   -16.376     4.662
 (0 - 3)     16.857     0.000     6.338    27.376
 (1 - 0)    -21.500     0.000   -32.019   -10.981
 (1 - 2)    -27.357     0.000   -37.876   -16.838
 (1 - 3)     -4.643     0.647   -15.162     5.876
 (2 - 0)      5.857     0.458    -4.662    16.376
 (2 - 1)     27.357     0.000    16.838    37.876
 (2 - 3)     22.714     0.000    12.195    33.233
 (3 - 0)    -16.857     0.000   -27.376    -6.338
 (3 - 1)      4.643     0.647    -5.876    15.162
 (3 - 2)    -22.714     0.000   -33.233   -12.195



**выводы**

Тест Тьюки показал, что на уровне доверия 95% значимых различий между средними весами тигров, получавших корма 1 и 3, а также 2 и 4, нет. Однако для остальных сочетаний групп выявлены значительные различия в весе.

Задание 12 (про белок)

In [4]:
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols

# Установим формат вывода для pandas
pd.options.display.float_format = '{:.3f}'.format

# Создадим словарь с данными
data = {
    'has_brood': [0]*14 + [1]*16,
    'age': [1]*5 + [2]*5 + [3]*5 + [1]*5 + [2]*5 + [3]*5,
    'weight': [440, 438, 429, 502, 602,
               892, 868, 855, 866, 932,
               1575, 849, 759, 1602, 1327,
               308, 328, 326, 326, 325,
               737, 798, 876, 810, 861,
               1001, 901, 958, 1032, 883]
}

# Преобразуем данные в DataFrame
df = pd.DataFrame(data)

# Определим модель с помощью OLS
formula = 'weight ~ has_brood + age + has_brood:age'
model = ols(formula, data=df).fit()

# Выполним ANOVA для оценки влияния факторов
anova_results = sm.stats.anova_lm(model, typ=1)

# Выведем результаты
print(anova_results)


                  df      sum_sq     mean_sq      F  PR(>F)
has_brood      1.000   63063.005   63063.005  1.888   0.181
age            1.000 2434837.659 2434837.659 72.881   0.000
has_brood:age  1.000    1219.187    1219.187  0.036   0.850
Residual      26.000  868618.949   33408.421    NaN     NaN


**Статистический вывод:**

На основе проведенного анализа, с использованием уровня значимости 5% (доверительный уровень 95%), можно сделать вывод о том, что имеется достаточное основание для отклонения нулевой гипотезы в пользу альтернативной.

**Содержательный вывод:**

Результаты исследования свидетельствуют о том, что наличие выводка и возраст белок имеют значительное влияние на их массу тела. В то же время, взаимодействие этих двух факторов не продемонстрировало статистической значимости.

In [ ]:
'''Ваш ответ на задание №12
- верно указана применяемая статистика и обоснование ее применения
- расчет выполнен через последовательные суммы квадратов
- получены расчетные значения статистики
- получены значения p_value
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''